I actually didn't test the returned value, because when None the code just crashes, so I decided to match the TypeError exception to fail the test nicely.

```
$ ./tests/runtests.py -n unit.pillar.test_saltclass
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 * Python Version: 2.7.13 (default, Sep 17 2017, 18:27:43) [GCC 4.2.1 Compatible FreeBSD Clang 4.0.0 (tags/RELEASE_400/final 297347)]
 * Transplanting configuration files to '/tmp/salt-tests-tmpdir/config'
 * Current Directory: /usr/home/coredumb/salt-test/salt
 * Test suite is running under PID 60749
 * Logging tests on /tmp/salt-runtests.log
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Starting unit.pillar.test_saltclass Tests
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.
----------------------------------------------------------------------
Ran 1 test in 0.043s

OK

============================  Overall Tests Report  ============================
***  No Problems Found While Running Tests  ************************************
================================================================================
OK (total=1, skipped=0, passed=1, failures=0, errors=0) 
============================  Overall Tests Report  ============================
```
Unpatched code will fail the test:
```
$ ./tests/runtests.py -n unit.pillar.test_saltclass
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 * Python Version: 2.7.13 (default, Sep 17 2017, 18:27:43) [GCC 4.2.1 Compatible FreeBSD Clang 4.0.0 (tags/RELEASE_400/final 297347)]
 * Transplanting configuration files to '/tmp/salt-tests-tmpdir/config'
 * Current Directory: /usr/home/coredumb/salt-test/salt
 * Test suite is running under PID 7667
 * Logging tests on /tmp/salt-runtests.log
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Starting unit.pillar.test_saltclass Tests
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
F
======================================================================
FAIL: test_succeeds (unit.pillar.test_saltclass.SaltclassPillarTestCase)
[CPU:100.0%|MEM:63.2%] 
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/home/coredumb/salt-test/salt/tests/unit/pillar/test_saltclass.py", line 51, in test_succeeds
    self._runner(ret)
  File "/usr/home/coredumb/salt-test/salt/tests/unit/pillar/test_saltclass.py", line 45, in _runner
    self.fail(err)
AssertionError: argument of type 'NoneType' is not iterable

----------------------------------------------------------------------
Ran 1 test in 0.046s

FAILED (failures=1)

============================  Overall Tests Report  ============================
*** unit.pillar.test_saltclass Tests  ******************************************
 --------  Failed Tests  -------------------------------------------------------
   -> unit.pillar.test_saltclass.SaltclassPillarTestCase.test_succeeds  ........
       Traceback (most recent call last):
         File "/usr/home/coredumb/salt-test/salt/tests/unit/pillar/test_saltclass.py", line 51, in test_succeeds
           self._runner(ret)
         File "/usr/home/coredumb/salt-test/salt/tests/unit/pillar/test_saltclass.py", line 45, in _runner
           self.fail(err)
       AssertionError: argument of type 'NoneType' is not iterable
   .............................................................................
 -------------------------------------------------------------------------------
================================================================================
FAILED (total=1, skipped=0, passed=0, failures=1, errors=0) 
============================  Overall Tests Report  ============================
```

Hope this is OK for you like this!

--- a/salt/utils/saltclass.py
+++ b/salt/utils/saltclass.py
@@ -104,11 +104,12 @@ def dict_search_and_replace(d, old, new, expanded):
 def find_value_to_expand(x, v):
     a = x
     for i in v[2:-1].split(':'):
+        if a is None:
+            return v
         if i in a:
             a = a.get(i)
         else:
-            a = v
-            return a
+            return v
     return a
 
 

--- a/tests/integration/files/saltclass/examples/classes/default/init.yml
+++ b/tests/integration/files/saltclass/examples/classes/default/init.yml
@@ -9,9 +9,11 @@ pillars:
   default:
     network:
       dns:
+{% if __grains__['os'] == 'should_never_match' %}
         srv1: 192.168.0.1
         srv2: 192.168.0.2
         domain: example.com
+{% endif %}
       ntp:
         srv1: 192.168.10.10
         srv2: 192.168.10.20

--- a/tests/integration/files/saltclass/examples/nodes/fake_id.yml
+++ b/tests/integration/files/saltclass/examples/nodes/fake_id.yml
@@ -1,6 +1,6 @@
 environment: base
 
 classes:
-{% for class in ['default'] %}
+{% for class in ['default', 'roles.app'] %}
   - {{ class }}
 {% endfor %}

--- a/tests/unit/pillar/test_saltclass.py
+++ b/tests/unit/pillar/test_saltclass.py
@@ -34,10 +34,17 @@ class SaltclassPillarTestCase(TestCase, LoaderModuleMockMixin):
                            }}
 
     def _runner(self, expected_ret):
-        full_ret = saltclass.ext_pillar(fake_minion_id, fake_pillar, fake_args)
-        parsed_ret = full_ret['__saltclass__']['classes']
+        full_ret = {}
+        parsed_ret = []
+        try:
+            full_ret = saltclass.ext_pillar(fake_minion_id, fake_pillar, fake_args)
+            parsed_ret = full_ret['__saltclass__']['classes']
+        # Fail the test if we hit our NoneType error
+        except TypeError as err:
+            self.fail(err)
+        # Else give the parsed content result
         self.assertListEqual(parsed_ret, expected_ret)
 
     def test_succeeds(self):
-        ret = ['default.users', 'default.motd', 'default']
+        ret = ['default.users', 'default.motd', 'default', 'roles.app']
         self._runner(ret)

